docs/tech-specs/basic-spec.md

```diff
@@ -6,6 +6,280 @@
+* Broker crashed attempting to deliver to some/all consumers but succeeded with others (if any)
+
+### Design considerations
+
+Based on the flow, non-trivial scenarios we want to derive at the following design considerations -
+
+* Acknowledgement should be tied to being able to store Message **and** sending the ACK signal, but not dependent on Producer received the ACK signal.
+* There should be a fail-safe way to ensure Acknowledged messages are attempted to deliver in case synchronous triggering of delivery process fails
+* Consumer may not be available, so we should have a retry policy
+
+### One level deeper
+
+So based on the above conversation lets elaborate the high-level flow to more concrete blocks and dive deep in them. The building blocks would be -
+
+1. Receive Message
+1. Start Delivery Process
+1. Attempt Individual Delivery
+
+Lets look into each of them.
+
+#### Receive Message
+
+* Start Transaction
+* Store Message
+* Commit Transaction
+* Send ACK
+* Trigger Delivery Process
+
+The acknowledgement transmission is intentionally not part of the transaction, as there is little bearing on the Producer actually receiving it. The worst case scenario will be Producer will resend the message.
+
+#### Start Delivery Process
+
+There is 2 entry points to the process.
+
+##### Triggered By Receipt Message
+
+* Retrieve the message
+* Retrieve the consumers
+* Start transaction
+* Create delivery status for each consumer
+* Mark message out-for-delivery
+* Commit transaction
+* Trigger independent asynchronous delivery process
+
+##### Triggered Fail-safe message
+
+* Retrieve all messages older than rational-delay
+* For each message use the Receipt Message trigger flow
+
+#### Attempt Individual Delivery
+
+For each message delivery we would have to follow the delivery lifecycle as designated by the _Message Delivery Status_. Here too, there will be 2 triggering functions -
+
+* Triggered by fail-safe mechanism
+  * Retrieve message deliveries in _Retry-Delivery_ with _next attempt timestamp_ past rational-delay
+  * For each message delivery run the _Triggered by Start Delivery Process_ process
+* Triggered by Start Delivery Process
+  * Update message delivery status to _Retry-In-Flight_ if the status is _Retry-Delivery_
+  * For delivery message retrieve consumer
+  * Attempt to deliver message
+  * Wait for the period of Delivery-Timeout
+    * If timed out then update status as _Retry-Delivery_ with exponentially backed off _next attempt timestamp_
+  * If work finishes before Timeout, mark status as _Delivered_
+  * If consumer failed to connect, update status as _Retry-Delivery_ with exponentially backed off _next attempt timestamp_
+  * If Max-Retries is reached and status is not _Delivered_ then mark delivery as _Dead_
+
+#### In a Sequence Diagram
+
+![Broker Message Lifecycle](./images/basic-spec/Broker_Message_Lifecycle.png)
+
+### Prioritized Dispatcher
```

Comment:
Priority is an integer and hence can have multiple values; but it is not `age` since there could be a future use case of `scheduled message` where age will be a factor; in this case priority is to signify what is picked up by worker first.